packages/start/src/client/createServerFn.ts

```diff
@@ -467,24 +543,77 @@ function serverFnBaseToMiddleware(
       validator: options.validator,
       validateClient: options.validateClient,
       client: async ({ next, sendContext, ...ctx }) => {
-        // Execute the extracted function
-        // but not before serializing the context
-        const res = await options.extractedFn?.({
+        const payload = {
           ...ctx,
           // switch the sendContext over to context
           context: sendContext,
-        } as any)
+        } as any
+
+        if (
+          options.static &&
+          process.env.NODE_ENV === 'production' &&
+          typeof document !== 'undefined'
+        ) {
+          const hash = jsonToFilenameSafeString(payload.data)
+          const url = getStaticCacheUrl(payload, hash)
+          let result: any = staticClientCache?.get(url)
+
+          if (!result) {
+            result = await fetch(url, {
+              method: 'GET',
+            })
+              .then((r) => r.text())
+              .then((d) => defaultTransformer.parse(d))
+
+            staticClientCache?.set(url, result)
+          }
+
+          if (result.error) {
+            throw result.error
+          }
+
+          return next(result.ctx)
+        }
+
+        // Execute the extracted function
+        // but not before serializing the context
+        const res = await options.extractedFn?.(payload)
 
         return next(res)
       },
       server: async ({ next, ...ctx }) => {
-        // Execute the server function
-        const result = await options.serverFn?.(ctx as any)
+        if (process.env.ROUTER !== 'client') {
+          // Execute the server function
+          const result = await options.serverFn?.(ctx as any)
+
+          return next({
+            result,
+          } as any)
+        }
 
-        return next({
-          result,
-        } as any)
+        throw new Error('Server function called from the client!')
       },
     },
   }
 }
+
+function jsonToFilenameSafeString(json: any) {
+  // Custom replacer to sort keys
+  const sortedKeysReplacer = (key: string, value: any) =>
+    value && typeof value === 'object' && !Array.isArray(value)
+      ? Object.keys(value)
+          .sort()
+          .reduce((acc: any, curr: string) => {
+            acc[curr] = value[curr]
+            return acc
+          }, {})
+      : value
+
+  // Convert JSON to string with sorted keys
+  const jsonString = JSON.stringify(json ?? '', sortedKeysReplacer)
+
+  // Replace characters invalid in filenames
+  return jsonString
```

Comment:
Perhaps this could be simplified with a `encodeURIComponent`?

Reply:
Worth a test. But not all url safe stuff is file-safe stuff

Reply:
Yeah, this change produced some invalid filenames that especially windows won't like. Keeping it for now